packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,236 @@
+        address owner
+    );
+    event DepositRewards(address indexed owner, uint256 amount);
+    event WithdrawRewards(address indexed owner, uint256 amount);
+    event SetWindowLock(address indexed owner, uint256 indexed windowIndex, bool locked);
+
+    modifier windowNotLocked(uint256 windowIndex) {
+        require(!merkleWindows[windowIndex].locked, "Window distributions locked");
+        _;
+    }
+
+    constructor(address _timerAddress) public Testable(_timerAddress) {}
+
+    /****************************
+     *
+     * Admin functions
+     *
+     ****************************/
+
+    // Set merkle root for the next available window index and seed allocations. Callable by owner of this
+    // contract. Importantly, we assume that the owner of this contract
+    // correctly chooses an amount `totalRewardsDistributed` that is sufficient
+    // to cover all claims within the `merkleRoot`. Otherwise, a race condition
+    // can be created. This situation can occur because we do not segregate reward balances by window,
+    // for code simplicity purposes.
+    //
+    // Example race situation:
+    //     - Window 1 Tree: Owner sets `totalRewardsDistributed=100` and insert proofs that give
+    //                      claimant A 50 tokens and claimant B 51 tokens. The owner has made an error
+    //                      by not setting the `totalRewardsDistributed` correctly to 101).
+    //     - Window 2 Tree: Owner sets `totalRewardsDistributed=1` and insert proofs that give
+    //                      claimant A 1 token. The owner correctly set `totalRewardsDistributed` this time.
+    //     - At this point contract owns 100 + 1 = 101 tokens. Now, imagine the following sequence:
+    //       (1) Claimant A claims 50 tokens for Window 1, contract now has 101 - 50 = 51 tokens.
+    //       (2) Claimant B claims 51 tokens for Window 1, contract now has 51 - 51 = 0 tokens.
+    //       (3) Claimant A tries to claim 1 token for Window 2 but fails because contract has 0 tokens.
+    //     - In summary, the contract owner created a race for step(2) and step(3) in which the first
+    //       claim would succeed and the second claim would fail, even though both claimants would expect
+    //       their claims to suceed.
+    function setWindowMerkleRoot(
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        uint256 indexToSeed = lastSeededIndex;
+        lastSeededIndex = indexToSeed.add(1);
+
+        _seedWindow(indexToSeed, totalRewardsDistributed, windowStart, rewardToken, merkleRoot);
+    }
+
+    // When `window.locked` is true, claims are blocked for that window.
+    function setWindowLock(uint256 windowIndex, bool lockValue) external nonReentrant() onlyOwner {
+        merkleWindows[windowIndex].locked = lockValue;
+        emit SetWindowLock(msg.sender, windowIndex, lockValue);
+    }
```

Comment:
What do you think the tradeoffs of the lock functionality vs delete?

1. Lock seems better if we're worried about accidentally deleting the wrong merkel root. I could see the argument that once these are uploaded, we should only be able to disable them, not remove them. Otherwise a fat-finger could cause us to have to regenerate and reupload a modified root that removes the claims that had already happened on the deleted root.

2. Delete seems better if we're worried about accidentally re-enabling a distribution that we had been intended to be locked permanently (and rewards withdrawn). This also makes claiming a _little_ cheaper.

Thoughts? I personally think 1 sounds a bit stronger.

Reply:
I don't see the owner accidentally re-enabling a locked distribution, but I see your point about. making the claim cheaper. However, deleting a root seems bit dangerous and I'd rather enforce the functionality that once state is stored on this contract it remains.

Reply:
Actually, I'm thinking about this some more and I think removing lock in favor of being able to destroy merkle roots seems best. It saves gas for claims because we remove the `lock` data, and I don't see why claims would ever need to be locked temporarily. If an error in the tree was made, then its best to just delete it (and stop all claims immediately), and re-upload (+ optionally withdraw any rewards).

@chrismaree wdyt?

Reply:
Deleting SGTM! Would be interested to hear @chrismaree's perspective!

Reply:
yup! I agree with your analysis. the scenario we'd want this would be an error in the payouts. we'd want to manually re-compute the tree with removed claimers and then add a new hash. in this scenario deleting and re-uploading is the best outcome.